### PR TITLE
feat: add log_level attribute to js_binary and run_js_binary

### PR DESCRIPTION
**Pull request by gregmagolan**
Throw out the `verbose` attribute which wasn't very useful and opting for a log_level. This will give users fine grained control over what logs js_binary produces. It will be especially useful for debugging problematic targets as you can turn debug logs on for a particular `js_binary`, `js_test` or `run_js_binary` target without busting the whole graph by using `--define`.

Example "debug" level output on a test target:

```
DEBUG: aspect_rules_js[js_test]: binary target BINDIR bazel-out/darwin-fastbuild/bin
DEBUG: aspect_rules_js[js_test]: binary target TARGET_CPU darwin
DEBUG: aspect_rules_js[js_test]: binary target COMPILATION_MODE fastbuild
DEBUG: aspect_rules_js[js_test]: binary target WORKSPACE aspect_rules_js
DEBUG: aspect_rules_js[js_test]: binary target BUILD_FILE_PATH example/BUILD.bazel
INFO: aspect_rules_js[js_test]: binary target @//example:test10_test
INFO: aspect_rules_js[js_test]: PWD /private/var/tmp/_bazel_greg/2dc7794fea861d84d8ddaf3b9040a436/sandbox/darwin-sandbox/1/execroot/aspect_rules_js/bazel-out/darwin-fastbuild/bin/example/_test10_test_launcher.sh.runfiles/aspect_rules_js
INFO: aspect_rules_js[js_test]: running /private/var/tmp/_bazel_greg/2dc7794fea861d84d8ddaf3b9040a436/sandbox/darwin-sandbox/1/execroot/aspect_rules_js/bazel-out/darwin-fastbuild/bin/example/_test10_test_launcher.sh.runfiles/aspect_rules_js/../nodejs_darwin_amd64/bin/nodejs/bin/node -- /private/var/tmp/_bazel_greg/2dc7794fea861d84d8ddaf3b9040a436/sandbox/darwin-sandbox/1/execroot/aspect_rules_js/bazel-out/darwin-fastbuild/bin/example/_test10_test_launcher.sh.runfiles/aspect_rules_js/example/case10.js dummy
```

Example "debug" level output on a build (run_js_binary) target:

```
DEBUG: aspect_rules_js[js_binary]: changing directory to BAZEL_BINDIR (root of Bazel output tree) bazel-out/darwin-fastbuild/bin
DEBUG: aspect_rules_js[js_binary]: BAZEL_BINDIR bazel-out/darwin-fastbuild/bin
DEBUG: aspect_rules_js[js_binary]: BAZEL_TARGET_CPU darwin
DEBUG: aspect_rules_js[js_binary]: BAZEL_COMPILATION_MODE fastbuild
DEBUG: aspect_rules_js[js_binary]: BAZEL_WORKSPACE aspect_rules_js
DEBUG: aspect_rules_js[js_binary]: BAZEL_BUILD_FILE_PATH example/BUILD.bazel
DEBUG: aspect_rules_js[js_binary]: BAZEL_INFO_FILE bazel-out/stable-status.txt
DEBUG: aspect_rules_js[js_binary]: BAZEL_VERSION_FILE bazel-out/volatile-status.txt
DEBUG: aspect_rules_js[js_binary]: binary target BINDIR bazel-out/darwin-opt-exec-2B5CBBC6/bin
DEBUG: aspect_rules_js[js_binary]: binary target TARGET_CPU darwin
DEBUG: aspect_rules_js[js_binary]: binary target COMPILATION_MODE opt
DEBUG: aspect_rules_js[js_binary]: binary target WORKSPACE aspect_rules_js
DEBUG: aspect_rules_js[js_binary]: binary target BUILD_FILE_PATH example/BUILD.bazel
INFO: aspect_rules_js[js_binary]: BAZEL_TARGET @//example:actual14_alt
INFO: aspect_rules_js[js_binary]: binary target @//example:rollup_bin
INFO: aspect_rules_js[js_binary]: PWD /private/var/tmp/_bazel_greg/2dc7794fea861d84d8ddaf3b9040a436/sandbox/darwin-sandbox/3/execroot/aspect_rules_js/bazel-out/darwin-fastbuild/bin
INFO: aspect_rules_js[js_binary]: running /private/var/tmp/_bazel_greg/2dc7794fea861d84d8ddaf3b9040a436/sandbox/darwin-sandbox/3/execroot/aspect_rules_js/bazel-out/darwin-opt-exec-2B5CBBC6/bin/example/_rollup_bin_launcher.sh.runfiles/aspect_rules_js/../nodejs_darwin_amd64/bin/nodejs/bin/node -- /private/var/tmp/_bazel_greg/2dc7794fea861d84d8ddaf3b9040a436/sandbox/darwin-sandbox/3/execroot/aspect_rules_js/bazel-out/darwin-opt-exec-2B5CBBC6/bin/example/_rollup_bin_launcher.sh.runfiles/aspect_rules_js/node_modules/.aspect_rules_js/rollup@2.70.2/node_modules/rollup/dist/bin/rollup --version
```

NB: these logs are supressed by default for run_js_binary targets on success since `silent_on_success` defaults to True.

Error message when BAZEL_BINDIR is not set is improved with this change:

FATAL: aspect_rules_js[js_binary]: BAZEL_BINDIR must be set in environment to the makevar $(BINDIR) in js_binary build actions (which run in the execroot) so that build actions can change directories to always run out of the root of the Bazel output tree. See https://docs.bazel.build/versions/main/be/make-variables.html#predefined_variables. This is automatically set by 'run_js_binary' (https://github.com/aspect-build/rules_js/blob/main/docs/run_js_binary.md) which is the recommended rule to use for using a js_binary as the tool of a build action. If this is not a build action you can set the BAZEL_BINDIR to '.' instead to supress this error. For more context on this design decision, please read the aspect_rules_js README https://github.com/aspect-build/rules_js#running-nodejs-programs.

### CODE REVIEW

File: js/private/js_binary.sh.tpl [hunk 1]
```diff
@@ -245,9 +357,12 @@ fi
 # ==============================================================================
 
 readonly child=$!
+# Bash does not forward termination signals to any child process when
+# running in docker so need to manually trap and forward the signals
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply js/private/js_binary.sh.tpl:361]: Cool. If there is a better approach lets do it as a follow-up. This code has been around for a while so I'd rather change it in a follow-up and not mix concerns with this PR. https://github.com/bazelbuild/rules_nodejs/blob/stable/internal/node/launcher.sh#L380

File: js/private/js_binary.sh.tpl [hunk 2]
```diff
@@ -9,8 +9,115 @@ set -o pipefail -o errexit -o nounset
...
+    STDERR_CAPTURE=$(mktemp)
+fi
+
 LOG_PREFIX="{{log_prefix_rule_set}}[{{log_prefix_rule}}]"
 
+function logf_stderr {
+    if [ "${STDERR_CAPTURE:-}" ]; then
+        # shellcheck disable=SC2059
+        printf "$@" >>"$STDERR_CAPTURE"
+        printf "\n" >>"$STDERR_CAPTURE"
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply js/private/js_binary.sh.tpl:30]: 🌮

File: js/private/js_binary.sh.tpl [hunk 3]
```diff
@@ -9,8 +9,115 @@ set -o pipefail -o errexit -o nounset
 
 {{envs}}
 
+# ==============================================================================
+# Prepare stdout capture, stderr capture && logging
+# ==============================================================================
+
+if [ "${JS_BINARY__STDOUT_OUTPUT_FILE:-}" ] || [ "${JS_BINARY__SILENT_ON_SUCCESS:-}" ]; then
+    STDOUT_CAPTURE=$(mktemp)
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply js/private/js_binary.sh.tpl:17]: played with this idea for a bit but with no luck. mktemp on osx doesn't appear to be doing what its manpage says. sticking with the system tmp directly seems safest. we do clean up after ourselves on exit and delete the tmp files created

File: js/private/test/BUILD.bazel [hunk 4]
```diff
@@ -28,12 +28,15 @@ js_binary(
...
 )
 
-# For consistency on different platform, replace the host nodejs platform to linux_amd64
+# Make sed replacements for consistency on different platform
 genrule(
     name = "shell_launcher_sed",
     srcs = [":shellcheck_launcher"],
     outs = ["shellcheck_launcher_sed.sh"],
-    cmd = "cat $(execpath :shellcheck_launcher) | sed \"s#%s#linux_amd64#\" > $@" % host.platform,
+    cmd = "cat $(execpath :shellcheck_launcher) | sed \"s#$(BINDIR)#bazel-out/k8-fastbuild/bin#\" | sed \"s#%s#linux_amd64#\" | sed \"s#\\\"%s\\\"#\\\"k8\\\"#\" > $@" % (
```
Comment (alexeagle) [inline js/private/test/BUILD.bazel:36]: "k8"??

File: js/run_js_binary.bzl [hunk 5]
```diff
@@ -106,41 +116,54 @@ def run_js_binary(
...
-        extra_env["JS_BINARY__CAPTURE_EXIT_CODE"] = "%s$(rootpath %s)" % (chdir_prefix, exit_code_out)
+        extra_env["JS_BINARY__EXIT_CODE_OUTPUT_FILE"] = "$(execpath {})".format(exit_code_out)
         extra_outs.append(exit_code_out)
 
     # Configure silent on success
     if silent_on_success:
         extra_env["JS_BINARY__SILENT_ON_SUCCESS"] = "1"
 
+    # Configure log_level if specified
+    if log_level:
```
Comment (alexeagle) [inline js/run_js_binary.bzl:152]: too bad this logic gets repeated so much

File: js/private/test/shellcheck_launcher.sh [hunk 6]
```diff
@@ -136,116 +257,110 @@ else
...
+    if [ "${BAZEL_BUILD_FILE_PATH:-}" ]; then
+        logf_debug "BAZEL_BUILD_FILE_PATH %s" "${BAZEL_BUILD_FILE_PATH:-}"
+    fi
+    if [ "${BAZEL_INFO_FILE:-}" ]; then
+        logf_debug "BAZEL_INFO_FILE %s" "${BAZEL_INFO_FILE:-}"
+    fi
+    if [ "${BAZEL_VERSION_FILE:-}" ]; then
+        logf_debug "BAZEL_VERSION_FILE %s" "${BAZEL_VERSION_FILE:-}"
+    fi
+    logf_debug "binary target BINDIR %s" "${JS_BINARY__BINDIR:-}"
```
Comment (alexeagle) [inline js/private/test/shellcheck_launcher.sh:332]: would it be shorter to logf_debug once with a multiline argument?

**Review by alexeagle — APPROVED** [review]

File: js/private/test/BUILD.bazel (hunk 4 above)
Comment (gregmagolan) [reply js/private/test/BUILD.bazel:36]: You mean `linux_amd64`? That one is correct as it comes from our canonical platform naming in aspect_bazel_lib and not from bazel.

File: js/run_js_binary.bzl (hunk 5 above)
Comment (gregmagolan) [reply js/run_js_binary.bzl:152]: Its not likely to change in the future and its just the two spots js_binary and its wrapper macro run_js_binary so should be just limited to those. Tho, I can abstract some more to reduce the number of duplicate lines of code...

Comment (gregmagolan) [reply js/run_js_binary.bzl:152]: Much cleaner
```
    # Configure log_level if specified
    if log_level:
        for env in _js_binary_envs_for_log_level(log_level):
            extra_env[env] = "1"
```
and
```
    for env in envs_for_log_level(ctx.attr.log_level):
        envs.append(_ENV_SET_IFF_NOT_SET.format(
            var = env,
            value = "1",
        ))
```

File: js/private/test/shellcheck_launcher.sh (hunk 6 above)
Comment (gregmagolan) [reply js/private/test/shellcheck_launcher.sh:332]: the ordering of the `%s` statements becomes unruly with a single call

File: js/private/test/BUILD.bazel (hunk 4 above)
Comment (alexeagle) [reply js/private/test/BUILD.bazel:36]: no, there's also a k8 near the end of the regex, seems fishy to be linux/processor specific

**Pull request merged**